Add support for Web Extension menus / contextMenu APIs.

xeenon opened the pull request:
#### b3a279cf875c838db8715ee56344d766e84e0791
<pre>
Add support for Web Extension menus / contextMenu APIs.
<a href="https://webkit.org/b/260158">https://webkit.org/b/260158</a>
<a href="https://rdar.apple.com/problem/114823238">rdar://problem/114823238</a>

Reviewed by Brian Weinstein.

* Source/WebKit/DerivedSources-input.xcfilelist:
* Source/WebKit/DerivedSources-output.xcfilelist:
* Source/WebKit/DerivedSources.make:
* Source/WebKit/Platform/cocoa/CocoaHelpers.mm:
(WebKit::parseJSON):
* Source/WebKit/Scripts/webkit/messages.py:
(types_that_cannot_be_forward_declared):
* Source/WebKit/Shared/Extensions/WebExtensionEventListenerType.h:
* Source/WebKit/Shared/Extensions/WebExtensionEventListenerType.serialization.in:
* Source/WebKit/Shared/Extensions/WebExtensionMenuItem.serialization.in: Added.
* Source/WebKit/Shared/Extensions/WebExtensionMenuItemContextParameters.h: Added.
* Source/WebKit/Shared/Extensions/WebExtensionMenuItemContextType.h: Added.
(WebKit::allWebExtensionMenuItemContextTypes):
* Source/WebKit/Shared/Extensions/WebExtensionMenuItemParameters.h: Added.
* Source/WebKit/Shared/Extensions/WebExtensionMenuItemType.h: Added.
(WebKit::isCheckedType):
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtension.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionAction.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionAction.mm:
(-[_WKWebExtensionAction menuItems]):
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionContext.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionContext.mm:
(-[_WKWebExtensionContext menuItemsForTab:]):
* Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPICommandsCocoa.mm:
(WebKit::WebExtensionContext::fireCommandEventIfNeeded):
(WebKit::WebExtensionContext::fireCommandChangedEventIfNeeded):
* Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIMenusCocoa.mm: Added.
(WebKit::isAncestorOrSelf):
(WebKit::WebExtensionContext::menusCreate):
(WebKit::WebExtensionContext::menusUpdate):
(WebKit::WebExtensionContext::menusRemove):
(WebKit::WebExtensionContext::menusRemoveAll):
(WebKit::WebExtensionContext::fireMenusClickedEventIfNeeded):
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionActionCocoa.mm:
(WebKit::WebExtensionAction::platformMenuItems const):
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
(WebKit::WebExtensionContext::command):
(WebKit::WebExtensionContext::platformMenuItems const):
(WebKit::WebExtensionContext::menuItem const):
(WebKit::WebExtensionContext::performMenuItem):
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionMenuItemCocoa.mm: Added.
(-[_WKWebExtensionMenuItem initWithTitle:handler:]):
(-[_WKWebExtensionMenuItem _performAction:]):
(WebKit::WebExtensionMenuItem::WebExtensionMenuItem):
(WebKit::WebExtensionMenuItem::minimalParameters const):
(WebKit::WebExtensionMenuItem::update):
(WebKit::WebExtensionMenuItem::matchingPlatformMenuItems):
(WebKit::WebExtensionMenuItem::platformMenuItem const):
(WebKit::WebExtensionMenuItem::icon const):
* Source/WebKit/UIProcess/Extensions/WebExtensionAction.h:
* Source/WebKit/UIProcess/Extensions/WebExtensionContext.h:
(WebKit::WebExtensionContext::mainMenuItems const):
* Source/WebKit/UIProcess/Extensions/WebExtensionContext.messages.in:
* Source/WebKit/UIProcess/Extensions/WebExtensionMenuItem.cpp: Added.
(WebKit::WebExtensionMenuItem::operator== const):
(WebKit::WebExtensionMenuItem::extensionContext const):
(WebKit::WebExtensionMenuItem::removeAmpersands):
(WebKit::WebExtensionMenuItem::matches const):
(WebKit::WebExtensionMenuItem::toggleCheckedIfNeeded):
(WebKit::WebExtensionMenuItem::addSubmenuItem):
(WebKit::WebExtensionMenuItem::removeSubmenuItem):
* Source/WebKit/UIProcess/Extensions/WebExtensionMenuItem.h: Added.
(WebKit::WebExtensionMenuItem::create):
(WebKit::WebExtensionMenuItem::type const):
(WebKit::WebExtensionMenuItem::identifier const):
(WebKit::WebExtensionMenuItem::title const):
(WebKit::WebExtensionMenuItem::command const):
(WebKit::WebExtensionMenuItem::isChecked const):
(WebKit::WebExtensionMenuItem::setChecked):
(WebKit::WebExtensionMenuItem::isEnabled const):
(WebKit::WebExtensionMenuItem::isVisible const):
(WebKit::WebExtensionMenuItem::documentPatterns const):
(WebKit::WebExtensionMenuItem::targetPatterns const):
(WebKit::WebExtensionMenuItem::contexts const):
(WebKit::WebExtensionMenuItem::parentMenuItem const):
(WebKit::WebExtensionMenuItem::submenuItems const):
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIActionCocoa.mm:
(WebKit::WebExtensionAPIAction::isValidDimensionKey):
(WebKit::isValidDimensionKey): Deleted.
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIMenusCocoa.mm: Added.
(WebKit::toMenuIdentifierWebAPI):
(WebKit::WebExtensionAPIMenus::parseCreateAndUpdateProperties):
(WebKit::WebExtensionAPIMenus::createMenu):
(WebKit::WebExtensionAPIMenus::update):
(WebKit::WebExtensionAPIMenus::remove):
(WebKit::WebExtensionAPIMenus::removeAll):
(WebKit::WebExtensionAPIMenus::onClicked):
(WebKit::WebExtensionContextProxy::dispatchMenusClickedEvent):
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPINamespaceCocoa.mm:
(WebKit::WebExtensionAPINamespace::menus):
* Source/WebKit/WebProcess/Extensions/API/WebExtensionAPIAction.h:
* Source/WebKit/WebProcess/Extensions/API/WebExtensionAPIMenus.h: Added.
(WebKit::WebExtensionAPIMenus::actionMenuTopLevelLimit const):
(WebKit::WebExtensionAPIMenus::clickHandlers const):
* Source/WebKit/WebProcess/Extensions/API/WebExtensionAPINamespace.h:
(WebKit::WebExtensionAPINamespace::contextMenus):
* Source/WebKit/WebProcess/Extensions/Bindings/Cocoa/JSWebExtensionWrapperCocoa.mm:
(WebKit::WebExtensionCallbackHandler::WebExtensionCallbackHandler):
* Source/WebKit/WebProcess/Extensions/Bindings/JSWebExtensionWrapper.h:
* Source/WebKit/WebProcess/Extensions/Interfaces/WebExtensionAPIMenus.idl: Added.
* Source/WebKit/WebProcess/Extensions/Interfaces/WebExtensionAPINamespace.idl:
* Source/WebKit/WebProcess/Extensions/WebExtensionContextProxy.h:
* Source/WebKit/WebProcess/Extensions/WebExtensionContextProxy.messages.in:
* Tools/TestWebKitAPI/SourcesCocoa.txt:
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPIMenus.mm: Added.
(TestWebKitAPI::TEST):
(TestWebKitAPI::performMenuItemAction):

Canonical link: <a href="https://commits.webkit.org/271259@main">https://commits.webkit.org/271259@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/6db10810f5da9bc6749f8cab738a35b2689c3d25

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/27831 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/6469 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/29079 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/30049 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/25407 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/28324 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/8428 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/3862 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/25164 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/28096 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/5214 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/23868 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/4518 "Passed tests") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/27991 "Passed tests") | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/4686 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/24890 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/30688 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/25420 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/25329 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/30868 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/4703 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/50/builds/2857 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/28770 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/6220 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/6678 "Built successfully and passed tests") | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/5149 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/5165 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

b-weinstein commented on Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIMenusCocoa.mm:
```diff
@@ -0,0 +1,67 @@
+
+void WebExtensionContext::menusCreate(const WebExtensionMenuItemParameters& parameters, CompletionHandler<void(std::optional<String>)>&& completionHandler)
+{
+    // FIXME: Implement.
```
Nit: Bug numbers for these? Or does it just come in the next commit?

xeenon replied:
Yep, next commit.

b-weinstein commented on Source/WebKit/UIProcess/Extensions/WebExtensionContext.h:
```diff
@@ -447,6 +447,12 @@ class WebExtensionContext : public API::ObjectImpl<API::Object::Type::WebExtensi
+    // Menus APIs
+    void menusCreate(const WebExtensionMenuItemParameters&, CompletionHandler<void(std::optional<String>)>&&);
+    void menusUpdate(const String& identifier, const WebExtensionMenuItemParameters&, CompletionHandler<void(std::optional<String>)>&&);
+    void menusRemove(const String& identifier, CompletionHandler<void(std::optional<String>)>&&);
```
Are these optional strings the error messages?

xeenon replied:
Yes

b-weinstein commented on Source/WebKit/UIProcess/Extensions/WebExtensionContext.messages.in:
```diff
@@ -55,6 +55,12 @@ messages -> WebExtensionContext {
+    // Menus APIs
+    MenusCreate(struct WebKit::WebExtensionMenuItemParameters parameters) -> (std::optional<String> error);
+    MenusUpdate(String identifier, struct WebKit::WebExtensionMenuItemParameters parameters) -> (std::optional<String> error);
+    MenusRemove(String identifier) -> (std::optional<String> error);
```
Ah - yes - it's more clear here.

b-weinstein commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIMenusCocoa.mm:
```diff
@@ -0,0 +1,358 @@
+    };
+
+    bool isSeparator = [objectForKey<NSString>(properties, typeKey) isEqualToString:separatorKey];
+    if (!validateDictionary(properties, @"properties", !isSeparator ? requiredKeys : nil, types, outExceptionString))
```
Super nit: I think `isSeparator ? nil : requiredKeys` is a bit more readable.

b-weinstein commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIMenusCocoa.mm:
```diff
@@ -0,0 +1,358 @@
+            if (([context isEqualToString:@"browser_action"] || [context isEqualToString:@"page_action"]) && extensionContext().supportsManifestVersion(3)) {
+                *outExceptionString = toErrorString(nil, contextsKey, @"'%@' is not a valid context", context);
+                return false;
+            }
```
Nice.

b-weinstein commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIMenusCocoa.mm:
```diff
@@ -0,0 +1,358 @@
+        parameters.title = title;
+    }
+
+    if (JSValue *clickCallback = properties[onclickKey]) {
```
Nit: `onclickKey`.

xeenon replied:
I have `onclickKey`.

b-weinstein replied:
I meant it should be `onClickKey`, but that's not right because it's the `onclick` event.

cdumez replied:
Not to be nit-picky for the event is normally `click`. `onclick` is the name of the event handler on the element :)

b-weinstein replied:
Actually - the event is `onClicked` - should it be `onClickedKey`?

b-weinstein commented on Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIMenusCocoa.mm:
```diff
@@ -32,36 +32,131 @@
+    if (potentialAncestorIdentifier == identifier)
+        return true;
+
+    RefPtr current = context.menuItem(identifier);
```
Does auto work here?

xeenon replied:
Yes, but `context.menuItem` returns a pointer, so this protects it while in scope.

Also even if `RefPtr` or `Ref` is returned, I've been told by @cdumez that we should be explicit and use `RefPtr` or `Ref` instead of `auto` as WebKit is adopting smart pointer aggressively and that makes the calcite clear to anyone doing a code audit or review.

b-weinstein commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionMenuItemCocoa.mm:
```diff
@@ -0,0 +1,311 @@
+        m_identifier = parameters.identifier;
+
+    if (parameters.parentIdentifier) {
+        RefPtr newParentMenuItem = m_extensionContext->menuItem(parameters.parentIdentifier.value());
```
Nit: `updatedParentMenuItem`? `new` makes it sound like a new one is being created.

b-weinstein commented on Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionMenuItemCocoa.mm:
```diff
@@ -0,0 +1,311 @@
+{
+    bool inRadioGroup = false;
+    bool groupChecked = false;
+    size_t count = 0;
```
Does this need `__block`? Or is that only for Objective-C blocks?

xeenon replied:
Only ObjC blocks, the compiler complains otherwise. The C++ lambda captures this as a reference with the `[&]` below.

b-weinstein commented on Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIMenusCocoa.mm:
```diff
@@ -353,6 +381,69 @@
+        info[srcURLKey] = contextParameters.sourceURL.string();
+
+    if (!contextParameters.linkURL.isNull())
+        info[linkURLKey] = contextParameters.linkURL.string();
```
If we are passing a string - could this be something like `linkURLStringKey`?

xeenon replied:
This is for the result being passed to JS, and string the only option.

kiaraarose commented on Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionContext.h:
```diff
@@ -525,6 +531,20 @@ WK_CLASS_AVAILABLE(macos(13.3), ios(16.4))
 - (void)performCommand:(_WKWebExtensionCommand *)command;
 
+/*!
+ @abstract Retrieves an array of menu item for a given tab.
```
nit: `items`

xeenon committed:
Add support for Web Extension menus / contextMenu APIs.
https://webkit.org/b/260158
rdar://problem/114823238

Reviewed by Brian Weinstein.

* Source/WebKit/DerivedSources-input.xcfilelist:
* Source/WebKit/DerivedSources-output.xcfilelist:
* Source/WebKit/DerivedSources.make:
* Source/WebKit/Platform/cocoa/CocoaHelpers.mm:
(WebKit::parseJSON):
* Source/WebKit/Scripts/webkit/messages.py:
(types_that_cannot_be_forward_declared):
* Source/WebKit/Shared/Extensions/WebExtensionEventListenerType.h:
* Source/WebKit/Shared/Extensions/WebExtensionEventListenerType.serialization.in:
* Source/WebKit/Shared/Extensions/WebExtensionMenuItem.serialization.in: Added.
* Source/WebKit/Shared/Extensions/WebExtensionMenuItemContextParameters.h: Added.
* Source/WebKit/Shared/Extensions/WebExtensionMenuItemContextType.h: Added.
(WebKit::allWebExtensionMenuItemContextTypes):
* Source/WebKit/Shared/Extensions/WebExtensionMenuItemParameters.h: Added.
* Source/WebKit/Shared/Extensions/WebExtensionMenuItemType.h: Added.
(WebKit::isCheckedType):
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtension.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionAction.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionAction.mm:
(-[_WKWebExtensionAction menuItems]):
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionContext.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionContext.mm:
(-[_WKWebExtensionContext menuItemsForTab:]):
* Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPICommandsCocoa.mm:
(WebKit::WebExtensionContext::fireCommandEventIfNeeded):
(WebKit::WebExtensionContext::fireCommandChangedEventIfNeeded):
* Source/WebKit/UIProcess/Extensions/Cocoa/API/WebExtensionContextAPIMenusCocoa.mm: Added.
(WebKit::isAncestorOrSelf):
(WebKit::WebExtensionContext::menusCreate):
(WebKit::WebExtensionContext::menusUpdate):
(WebKit::WebExtensionContext::menusRemove):
(WebKit::WebExtensionContext::menusRemoveAll):
(WebKit::WebExtensionContext::fireMenusClickedEventIfNeeded):
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionActionCocoa.mm:
(WebKit::WebExtensionAction::platformMenuItems const):
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionContextCocoa.mm:
(WebKit::WebExtensionContext::command):
(WebKit::WebExtensionContext::platformMenuItems const):
(WebKit::WebExtensionContext::menuItem const):
(WebKit::WebExtensionContext::performMenuItem):
* Source/WebKit/UIProcess/Extensions/Cocoa/WebExtensionMenuItemCocoa.mm: Added.
(-[_WKWebExtensionMenuItem initWithTitle:handler:]):
(-[_WKWebExtensionMenuItem _performAction:]):
(WebKit::WebExtensionMenuItem::WebExtensionMenuItem):
(WebKit::WebExtensionMenuItem::minimalParameters const):
(WebKit::WebExtensionMenuItem::update):
(WebKit::WebExtensionMenuItem::matchingPlatformMenuItems):
(WebKit::WebExtensionMenuItem::platformMenuItem const):
(WebKit::WebExtensionMenuItem::icon const):
* Source/WebKit/UIProcess/Extensions/WebExtensionAction.h:
* Source/WebKit/UIProcess/Extensions/WebExtensionContext.h:
(WebKit::WebExtensionContext::mainMenuItems const):
* Source/WebKit/UIProcess/Extensions/WebExtensionContext.messages.in:
* Source/WebKit/UIProcess/Extensions/WebExtensionMenuItem.cpp: Added.
(WebKit::WebExtensionMenuItem::operator== const):
(WebKit::WebExtensionMenuItem::extensionContext const):
(WebKit::WebExtensionMenuItem::removeAmpersands):
(WebKit::WebExtensionMenuItem::matches const):
(WebKit::WebExtensionMenuItem::toggleCheckedIfNeeded):
(WebKit::WebExtensionMenuItem::addSubmenuItem):
(WebKit::WebExtensionMenuItem::removeSubmenuItem):
* Source/WebKit/UIProcess/Extensions/WebExtensionMenuItem.h: Added.
(WebKit::WebExtensionMenuItem::create):
(WebKit::WebExtensionMenuItem::type const):
(WebKit::WebExtensionMenuItem::identifier const):
(WebKit::WebExtensionMenuItem::title const):
(WebKit::WebExtensionMenuItem::command const):
(WebKit::WebExtensionMenuItem::isChecked const):
(WebKit::WebExtensionMenuItem::setChecked):
(WebKit::WebExtensionMenuItem::isEnabled const):
(WebKit::WebExtensionMenuItem::isVisible const):
(WebKit::WebExtensionMenuItem::documentPatterns const):
(WebKit::WebExtensionMenuItem::targetPatterns const):
(WebKit::WebExtensionMenuItem::contexts const):
(WebKit::WebExtensionMenuItem::parentMenuItem const):
(WebKit::WebExtensionMenuItem::submenuItems const):
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIActionCocoa.mm:
(WebKit::WebExtensionAPIAction::isValidDimensionKey):
(WebKit::isValidDimensionKey): Deleted.
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIMenusCocoa.mm: Added.
(WebKit::toMenuIdentifierWebAPI):
(WebKit::WebExtensionAPIMenus::parseCreateAndUpdateProperties):
(WebKit::WebExtensionAPIMenus::createMenu):
(WebKit::WebExtensionAPIMenus::update):
(WebKit::WebExtensionAPIMenus::remove):
(WebKit::WebExtensionAPIMenus::removeAll):
(WebKit::WebExtensionAPIMenus::onClicked):
(WebKit::WebExtensionContextProxy::dispatchMenusClickedEvent):
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPINamespaceCocoa.mm:
(WebKit::WebExtensionAPINamespace::menus):
* Source/WebKit/WebProcess/Extensions/API/WebExtensionAPIAction.h:
* Source/WebKit/WebProcess/Extensions/API/WebExtensionAPIMenus.h: Added.
(WebKit::WebExtensionAPIMenus::actionMenuTopLevelLimit const):
(WebKit::WebExtensionAPIMenus::clickHandlers const):
* Source/WebKit/WebProcess/Extensions/API/WebExtensionAPINamespace.h:
(WebKit::WebExtensionAPINamespace::contextMenus):
* Source/WebKit/WebProcess/Extensions/Bindings/Cocoa/JSWebExtensionWrapperCocoa.mm:
(WebKit::WebExtensionCallbackHandler::WebExtensionCallbackHandler):
* Source/WebKit/WebProcess/Extensions/Bindings/JSWebExtensionWrapper.h:
* Source/WebKit/WebProcess/Extensions/Interfaces/WebExtensionAPIMenus.idl: Added.
* Source/WebKit/WebProcess/Extensions/Interfaces/WebExtensionAPINamespace.idl:
* Source/WebKit/WebProcess/Extensions/WebExtensionContextProxy.h:
* Source/WebKit/WebProcess/Extensions/WebExtensionContextProxy.messages.in:
* Tools/TestWebKitAPI/SourcesCocoa.txt:
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKWebExtensionAPIMenus.mm: Added.
(TestWebKitAPI::TEST):
(TestWebKitAPI::performMenuItemAction):

Canonical link: https://commits.webkit.org/271259@main
